### PR TITLE
Update pullspecs from provided CSV modifications

### PR DESCRIPTION
**Pull request by MartinBasti**
Update pullspecs in operator CSV file accordingly to provided JSON file
url with operator CSV modifications

* CLOUDBLD-3976

Signed-off-by: Martin Bašti <mbasti@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [n/a] JSON/YAML configuration changes are updated in the relevant schema
- [n/a] Changes to metadata also update the documentation for the metadata
- [n/a] Pull request has a link to an osbs-docs PR for user documentation updates
- [n/a] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 1]
```diff
@@ -66,6 +67,30 @@ def __init__(self, tasker, workflow, replacement_pullspecs=None,
...
+            return self._operator_csv_modifications
+
+        if not self.operator_csv_modifications_url:
+            return None
+
+        session = get_retrying_requests_session()
+
+        self.log.info("Fetching operator CSV modifications data from %s", self.operator_csv_modifications_url)
+        resp = session.get(self.operator_csv_modifications_url)
+        # TODO reraise with more user friendly error
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:86]: :eyes:

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 2]
```diff
@@ -66,6 +67,30 @@ def __init__(self, tasker, workflow, replacement_pullspecs=None,
...
+        resp.raise_for_status()
+        # TODO potentionally invalid JSON, reraise with user friendly error
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:88]: :eyes:

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 3]
```diff
@@ -131,6 +156,7 @@ def run_in_orchestrator(self):
         pullspecs = self._get_pullspecs(operator_manifest.csv, should_skip)
 
         if operator_manifest.csv.has_related_images() or should_skip:
+            # TODO do we want to support CSV modifications also in this case?
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:159]: :eyes:

Comment (rcerven) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:162]: well freshmaker might rebuild any bundle I'd say, so even when related images where already specified, freshmaker wants rebuild with new pullspecs, WDYT @lcarva

File: atomic_reactor/plugins/pre_pin_operator_digest.py [hunk 4]
```diff
@@ -262,6 +291,90 @@ def _get_pullspecs(self, operator_csv, skip_all):
...
+
+        :rtype: list[dict[str, ImageName or bool]]
+        :raises RuntimeError: if provided CSV modification doesn't contain all
+                              required pullspecs or contain different ones
+        """
+        mod_pullspec_repl = self.operator_csv_modifications.get('pullspec_replacements', [])
+
+        # check if modification data contains all required pullspecs
+        pullspecs_set = set(pullspecs)
+        mod_pullspecs_set = set((ImageName.parse(p['original']) for p in mod_pullspec_repl))
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_pin_operator_digest.py:348]: you should also verify that in provided replacements aren't duplicate 'original' entries with different other values

YOU might ignore duplicates IF all keys for duplicate entry are the same,
but I'd still say there shouldn't be any duplicates in provided replacements

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 3 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:170]: I'd address this in separate PR to unblock rest of the work

Comment (mprahl) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:170]: Freshmaker will not rebuild bundle images where the relatedImages section was not filled in by OSBS. It currently checks for the value of `True` for the `created_by_osbs` key.

Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:170]: Thanks! Added explicit check that this is not allowed (yet?)

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 4 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:357]: I'd address this in PR where JSONschema validation will be added

File: atomic_reactor/plugins/pre_pin_operator_digest.py (hunk 3 above)
Comment (rcerven) [reply atomic_reactor/plugins/pre_pin_operator_digest.py:170]: well so if it isn't supported we should fail here in that case

Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pin_operator_digest.py]: It works better when the one push local changes :D

**Review by rcerven — APPROVED** [review]

**Review by mkosiarc — APPROVED** [review]

**Review by ben-alkov — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]

**Pull request merged**